### PR TITLE
feat(charts): add a per-chart minimum zoom level

### PR DESCRIPTION
**Pull request by mairas**
## What problem does this solve?

Chart sets covering the same water at different scales are drawn on top of each
other, and a large-scale sheet is clutter until you are zoomed in far enough to
want it. Every hydrographic office publishes raster charts in overlapping scale
bands — NOAA's RNC series is Overview / General / Coastal / Approach / Harbor,
Traficom's Finnish sets are Yleiskartat / Merikarttasarjat / Rannikkokartat /
Satamakartat. Freeboard already picks the right band automatically for S-57,
because usage band is in the standard; raster has no equivalent, so the only fix
was ticking and unticking charts by hand while navigating.

## How does it work?

A chart can be given a lowest zoom level to appear at. It is a local preference
(`config.selections.chartDisplayMinZoom`), stored per chart id alongside the
existing per-chart opacity and image adjustment, and stripped from the chart the
properties dialog sends back to the server.

`resolveLayerZoomRange()` in `chart-utils.ts` combines the chart's declared range,
the user's minimum and the global over-zoom setting in one place; the raster, WMS,
WMTS and TileJSON layers use it. The minimum only ever restricts — a chart is
never drawn below the zoom its own tiles start at — and the top of the range is
untouched, so "Keep tiles visible on max zoom" behaves exactly as before. Charts
with no minimum resolve exactly as they did.

One change reaches outside charts: the status-bar zoom readout now truncates to a
tenth instead of rounding. Rounding reported `Zm: 14.0` from z13.95 up, so a chart
set to appear from z14 was legitimately off the map at a zoom the app was calling
14.0. Truncating means a level shown is always one the view has reached, and the
readout, the dialog's capture button and the layer bound all now use the same
convention.

## What changes for the user?

You can give each chart in an overlapping set the zoom it starts earning its place
at — coastal from z9, boating from z12, harbour from z14 — and each appears as you
zoom in, over the top of the coarser charts already drawn. Which one you see where
several overlap is still decided by chart order.

The control is in **Charts**, on each raster chart's row: the **minimum zoom**
button (a downward arrow onto a baseline), next to opacity and image adjustment.
It opens a small modeless dialog beside the chart list, which stays open, because
balancing an overlapping set means moving between charts. Type a level, or press
the target button to take the map's current zoom. The dialog shows the chart's own
data range for reference, and APPLY commits; closing it without applying puts the
old value back.

A chart carrying a minimum shows `from z12` under its name in the list, and adds
`— hidden at this zoom` while it is not being drawn — so a ticked chart that is
not on screen says why rather than looking broken. Charts you never set a minimum
on behave exactly as before.

The zoom readout in the status bar now counts up rather than rounding: a view a
hair under z14 reads `Zm: 13.9` where it used to read `Zm: 14.0`. It never claims
a level the map has not reached, which is what makes "set 14, zoom to 14, see the
chart" work.

## Known limitation

`chart.list` in the Plotter Extensions `charts` capability still reports
`visible: true` for a chart hidden by its minimum, and its `minZoom` is the
declared tile minimum rather than the effective one. That is a documented
host-agnostic contract, so it is not changed here — raised as
[joelkoz/signalk-plotterext-bus#2](https://github.com/joelkoz/signalk-plotterext-bus/issues/2)
with the options laid out.

## How was this tested?

Unit tests cover the resolver's precedence and bounds, config persistence and
injection, the dialog's parse/validate/label helpers, the dialog lifecycle (apply,
clear, revert, replacement), and the zoom-display truncation — including the case
that every zoom the readout shows as `14.0` draws a chart configured for z14.

Verified against a live Signal K server with the four Traficom sets: sweeping
z7.6→z14.8 and back draws exactly one chart at every sample with clean handovers
at each configured level, and a chart set to 14 now appears at the moment the
readout reaches 14.0.

CI green across the matrix; `npm run build:web` and `npm run test:ci` clean
locally too, bar two specs that fail the same way on an untouched `origin/master`
here and pass in CI.

## Checklist

- [x] One logical change; version numbers untouched.
- [x] `features/` untouched — the user-facing change is described above instead.
- [x] Skimmed [`DEV-LESSONS-LEARNED.md`](../docs/freeboard/DEV-LESSONS-LEARNED.md) for traps relevant to this change.
- [x] Tests added/updated for new behaviour and `npm run test:ci` passes.

---

**Second of three.** Stacked on #678, so its two commits appear here too; #680 is
stacked on this one. Merge order **#678 → #679 → #680**.

### CODE REVIEW

**Comment by joelkoz** [comment]
@coderabbitai review

File: features/changelog.json [hunk 1]
```diff
@@ -65,5 +65,7 @@
...
   { "feature": null, "kind": "skip", "date": "2026-05-23", "title": "perf(app): lazy-load dialog components on demand", "reason": "performance work — no change to what the user can do" },
   { "feature": null, "kind": "skip", "date": "2026-05-23", "title": "perf(s57): cache derived text-style config across features", "reason": "performance work — no change to what the user can do" },
-  { "feature": null, "kind": "skip", "date": "2026-04-04", "title": "perf: defer S57 chart symbols loading until first use", "reason": "performance work — no change to what the user can do" }
+  { "feature": null, "kind": "skip", "date": "2026-04-04", "title": "perf: defer S57 chart symbols loading until first use", "reason": "performance work — no change to what the user can do" },
+  { "feature": "chart-image-adjustment", "pr": 678, "kind": "enhanced", "date": "2026-08-07", "title": "fix(charts): make the adjustment palette usable beside the chart list" },
+  { "feature": "chart-minimum-zoom", "pr": 679, "kind": "new", "date": "2026-08-07", "title": "feat(charts): add a per-chart minimum zoom level" }
```
[the remark this answers is not in the document]
Comment (joelkoz) [reply features/changelog.json]: You have flagged this false positive on a number of PRs I have looked at.  The `features/changelog.json` is expected to have multiple entries with the same "pr" value. It would only be a true duplicate if the `feature` and `pr` properties of two rows were non-null and identical.

**Comment by joelkoz** [comment]
This is the strongest of the three, and I want to make sure you know it's less
niche than your own framing suggests.

You motivate it with the four Traficom sets, which reads as an archipelago
problem. It isn't — **every hydrographic office publishes raster charts in
overlapping scale bands.** NOAA's RNC series is Overview / General / Coastal /
Approach / Harbor, which is exactly the structure you describe. Freeboard already
picks the right band automatically for S-57 because usage band is in the
standard; raster has no equivalent, and that gap is what this fills. So it lands
for US coastal users too, including me. It's also the only one of the three PRs
with value *underway* rather than at the dock — your own line about "ticking and
unticking charts by hand while navigating" is the tell.

I wired the branch up to my dev server and ran it against my own NOAA raster
stack in the Florida Keys. Three things, the first of which is a real bug.

---

**1. A chart set to minimum zoom 14 does not appear at an indicated zoom of 14.**

I set the NOAA Hi-Res Relief layer to a minimum of 14, zoomed until the status
bar read `Zm: 14.0`, and the chart was not drawn. One more trackpad tick and it
appeared. Measured through the running app rather than guessed:

| | |
|---|---|
| actual view zoom | `13.975734194810341` |
| status bar (`mapZoomLevel().toFixed(1)`) | `Zm: 14.0` |
| layer bound (`displayMinZoom - MIN_ZOOM_EPSILON`) | `13.999999` |
| OL `inView`: `13.9757 > 13.999999` | false → **hidden** |

The status bar **rounds** to one decimal, so anything from just above 13.95 up to
14.0 reads as `14.0` while sitting below the bound. That is a dead band roughly
0.05 wide where the app tells the user they are at 14 and the chart stays hidden.
I had to cross 0.0243 zoom levels that the readout cannot show me.

You clearly knew about OL's minimum being exclusive — the epsilon is there for
exactly that reason, and the comment explaining it is right. The problem is that
`1e-6` is sized for exact float equality, a case continuous zoom essentially
never produces, while the granularity a user can actually see and set is `0.1`.
The epsilon is about 20,000× too small to cover the rounding in the readout.

I suspect your z7.6→z14.8 sweep missed it for the same reason: sampled at
intervals, a 0.05-wide band falls between the samples every time.

There's a third notion of "current zoom" in play too. At my 13.9757 the dialog's
**use current map zoom** button captures `Math.floor(13.9757 * 10) / 10` = `13.9`,
while the status bar beside it says `14.0`. So the display rounds, the capture
floors, and the comparison is exact — three answers to the same question.

The fix I'd suggest is to compare at the precision the user can see: draw the
chart when `Math.round(zoom * 10) / 10 >= displayMinZoom`, so what the status bar
reads is what gets tested. Failing that, widen the bound to half a display step
(`displayMinZoom - 0.05`) and round the capture button so it agrees with the
readout. Either way I'd want a test pinning the case, since it's the first thing
a user does with this feature: set a minimum, zoom to that number, expect the
chart.

---

**2. The icon.** These are Material Icons ligature names, the strings that go in
`<mat-icon>` — the control currently uses **`open_in_full`**, the diagonal
double-headed arrow that normally means expand-to-fullscreen.

I'd lean toward **`vertical_align_bottom`** (an arrow pointing down onto a
baseline), which reads as "lower bound" and is closer to what the setting
actually is. That said, `open_in_full` is defensible and I'm not going to spend
your time on it — your call between those two. The one I'd avoid is `zoom_in`
(the magnifier with a `+`): it says "zoom the map in," which is an action, not a
threshold, and there's a real chance of it being read as a button that changes
the view.

Worth noting the row's action bar is now at five or six icon buttons in a 315px
panel. It fits, but there's no room for a seventh, so "one more per-chart control
= one more icon" doesn't scale much further.

---

**3. Please open the plotter-extensions issue.** You were right to leave
`chart.list` alone here — that contract is host-agnostic and changing it inside a
Freeboard PR would have been the wrong move. Take me up on the follow-up though.
I confirmed the divergence live while testing: with the NOAA layer's effective
minimum set to 14, `chart.list` still reports `minZoom: 8` and `visible: true`
for it. Freeboard is the only known implementation of the `charts` capability at
the moment, so nothing is broken in the wild today — but that is exactly why it
should be settled deliberately now, while the contract is still cheap to change,
rather than after a second host reads it the other way.

Open the issue with the problem stated and both options on the table (report the
effective minimum, or add a `displayMinZoom` field to `ChartLayer`); I'll go
through the API contract and decide which way I want to take it. No need for you
to pick — just get it written down somewhere other than this thread.

---

On CodeRabbit's review: I've already dealt with the 🔴 Critical on
`features/changelog.json` — it was a false positive and it has recorded a
learning note, so it should stop recurring across the stack. The one worth taking
is `cleanDisplayMinZoomConfig()` not persisting its deletions — the entries do
come back after a restart. The `isBoundedOut` finding is the same one you
answered on #680, and I'd just point it at that thread; your reasoning there
convinced me. The `@let` nitpick is worth taking on its merits, since
`minZoomLabelFor(r)` runs twice per row on every change-detection pass.

On merge order: I've laid out the plan on #680, but the short version is that I
want to land **#678 → #679 → #680** as three separate merges rather than collapse
them, so each one needs to stand on its own. Three commits currently living only
on `feat/inline-chart-reorder` belong on this branch — `a0054f2b` (minimum zoom
input takes tenths), `ce646b58` (spinner steps by whole levels) and `12ad81f2`
(the `displayMinZoom` clone test, which CodeRabbit filed against #678 even though
the field doesn't exist there). As it stands this branch still has the step bug
you already solved. The point 1 fix above lands here too.

**Comment by mairas** [comment]
All four points addressed. The stack is rebased so this PR stands on its own —
`12ad81f2`, `a0054f2b` and `ce646b58` now live here (as `890f7e3b`, `8d3a705b`,
`e55231be`), so the step bug is fixed on this branch rather than three PRs
downstream.

---

**1. The dead band — fixed, but not the way you proposed.** You reproduced it
exactly and the diagnosis is right: the readout rounds, the capture floors, the
comparison is exact, and `1e-6` is sized for a float equality that continuous
zoom never produces.

Where I ended up differing is *which* of the three to move. Your fix widens the
comparison to meet the readout — draw when `Math.round(zoom * 10) / 10 >=
displayMinZoom`, which is a layer bound of `displayMinZoom - 0.05`. That closes
the band, but it does it by quietly turning a minimum of 14 into 13.95. A user
who types 14 has said something specific, and the chart would then be drawn at a
zoom below it.

The other direction closes the same band without that: **truncate the readout
instead of rounding it** (`73978cd3`). `13.9757` now reads `Zm: 13.9`, so the app
stops claiming a level the view has not reached, and a chart set to 14 appears
exactly when the readout says 14.0. The capture button already truncated, so all
three notions of "current zoom" collapse to one convention and the bound stays
exact.

Two things fell out of it:

- The truncation is now one helper (`src/app/lib/zoom-display.ts`) shared by the
  readout and the capture button, rather than the same expression written twice
  and free to drift.
- `Math.floor(zoom * 10) / 10` is wrong at a tenth: `12.3 * 10` is
  `122.99999999999999`, so the capture button turned a map at exactly z12.3 into a
  bound of `12.2`. The helper settles the product first.

The test you asked for is in `zoom-display.spec.ts`: every zoom the readout shows
as `14.0` must draw a chart configured for z14, asserted against the real
`resolveLayerZoomRange()` output. It fails if the readout ever goes back to
rounding.

**2. The icon — taken** (`12e95cb3`). `vertical_align_bottom` on all three sites:
the row's action button, the small glyph on the `from z14` label, and the dialog
header. Agreed on `zoom_in`; an action glyph on a threshold setting would read as
a button that moves the view.

**3. The issue is open**:
[joelkoz/signalk-plotterext-bus#2](https://github.com/joelkoz/signalk-plotterext-bus/issues/2).
Filed there rather than on Freeboard because the README calls the wire format the
contract and that is where `ChartLayer` is defined. Both your options are on the
table, unpicked, plus a third shape (a separate `rendered`/`onScreen` boolean) so
it is at least named while you decide.

**4. CodeRabbit.**

- `cleanDisplayMinZoomConfig()` — real, and it was worse than filed: the two
  pre-existing siblings, `cleanOpacityConfig()` and
  `cleanImageAdjustmentConfig()`, never persisted either, so orphaned opacity and
  image-adjustment entries have been coming back after a restart too. Fixing only
  the new one would have shipped three methods with two behaviours, so all three
  now share one helper and the caller writes once (`10e5d163`). Flagging the
  scope explicitly since two of those three are not mine.
- `@let` — taken (`63b7224a`).
- `isBoundedOut` — as you suggested, pointing at the #680 thread.
- `changelog.json` — moot; `features/` is reverted across the stack (see below).

---

**`features/` reverted.** #682 landed while I was working on this, so the whole
stack now follows `CONTRIBUTING.md` rather than just your comment: no
`chart-minimum-zoom.md`, no ledger rows, no assets. The user-facing description
lives in the PR body.

**Verification.** `npm run build:web` clean. `npm run test:ci`: 564 passing, 8
failing — all of them `whats-new.spec.ts` and `plotterext.embedding-host.spec.ts`,
which fail identically on a clean `origin/master` checkout. Nothing here touches
either.

**Comment by mairas** [comment]
Rebased onto the updated #678. Two commits here.

**The dialog fixture typing** (`57894a92`). CodeRabbit's finding was narrower
than the problem: the fixture's `any` was one of two copies of the same data
contract, the other being the inline object literal in the dialog's
`MAT_DIALOG_DATA` injection. Either could drift from the service that fills it
and nothing would say so — the tests would keep asserting against a shape the
dialog no longer reads. It is now `ChartMinZoomDialogData`, named once and used
by both. The minimum-zoom call site also got the explicit result type its
image-adjustment counterpart now carries on #678; inferred, `R` was `unknown`
and the `ChartMinZoomDialogResult` on the `afterClosed` callback was an
assertion rather than a check.

**The settings cleanup — I went further than the fix you endorsed** (`8dd5a30b`),
because persisting it made a second finding of CodeRabbit's true, and that one
was the more serious of the two.

`initItems()` treated every id missing from a listing as a chart that no longer
exists. A listing is not that statement: a provider that is down for the session
takes its charts out of it and brings them back later — the same case
`39efe5c5` already had to keep out of the stored chart order on #680. Before
`a4ee31f7` the deletions were in memory only, so a restart brought the settings
back. Adding the write turned a session-long annoyance into permanent loss: open
the chart list once while a provider is down and the opacity, image adjustment
and minimum zoom tuned for those charts are gone for good.

So the cleanup moved to where the app actually knows a chart is gone rather than
merely absent — `deleteChart()`, after the server confirms the delete. A refresh
no longer touches the settings at all.

The trade I made: charts deleted from another client leave their entries behind,
costing a few config keys nobody sees. Against that, a minimum zoom balanced
across a chart set is unrecoverable if it is dropped while the chart is only
away. If you would rather have the sweep back, the place for it is a listing
known to be complete, which nothing in the response currently tells us.

Tests are in `resources-delete-chart.spec.ts` (confirmed delete drops the
settings, a declined or failed delete does not) and
`chartlist-refresh.spec.ts` (a short listing keeps them, and writes nothing).
Both refresh assertions fail against the previous behaviour — that pair is the
bug reproduction.

**Verification.** `npm run build:web` clean. `npm run test:ci`: 561 passing, 0
failing.

---

**On the test-failure note — you are right that it is not a defect, but the
cause is not what you found.** I ran the full `npm run test:ci`, not vitest
against selected files, and it reproduces exactly: **8 failed | 564 passed**,
`whats-new.spec.ts` and `plotterext.embedding-host.spec.ts`, both at
`localStorage.clear()`.

The error is `TypeError: Cannot read properties of undefined (reading 'clear')`
rather than `localStorage is not defined`, which is the tell. I am on **Node
26.3.1**; CI pins `24.x`. Node ≥22 exposes an experimental `localStorage`
global that is `undefined` unless `--localstorage-file` is passed:

```
$ node -e "console.log(typeof localStorage, localStorage)"
undefined undefined
(node:5320) ExperimentalWarning: localStorage is not available because --localstorage-file was not provided
```

It shadows the jsdom one, and those two specs are the only specs under `src/`
that touch `localStorage` directly. Re-running the same wrapper with
`NODE_OPTIONS=--localstorage-file=…` and changing nothing else turns
`whats-new.spec.ts` green, 9/9 — which is how every count quoted above was
produced.

So both observations were right and neither was the whole picture: your run is
green because of the Node it ran on, mine was red because of mine. Worth
recording as an environment lesson, since Node 26 is what a contributor
installing "latest" gets today.

@coderabbitai you recorded a learning from this thread saying not to treat
failures from direct Vitest runs against selected spec files as baseline
failures, on the grounds that such a run bypasses the project configuration.
That diagnosis is wrong — the failures reproduce under the full
`npm run test:ci`. Please replace it: the cause is Node ≥22 exposing a
`localStorage` global that is `undefined` without `--localstorage-file`, which
shadows the jsdom one on Node 26. `npm run test:ci` is still the right command;
what it needs is a Node version CI covers, or that flag.

**Comment by mairas** [comment]
Amended and force-pushed: `resources-delete-chart.spec.ts` failed `format:check` on every platform. The settings-cleanup commit is now `9bee648d` (was `8dd5a30b`) — same content, Prettier-clean. Everything green apart from the armv7 leg.

**Comment by joelkoz** [comment]
You're right and I was wrong — thanks for running it down properly.

My diagnosis was that you'd run vitest against selected spec files and skipped the
project config. That's what *I* did, an hour before writing to you, and I generalised
from my own bad run to yours without checking that the failures matched. They didn't,
and the error text was the tell I walked past: mine was `localStorage is not defined`,
yours was `Cannot read properties of undefined (reading 'clear')`. Different failure,
different cause.

Your explanation holds and I've reproduced the reasoning: Node ≥22 exposes an
experimental `localStorage` global that stays `undefined` unless `--localstorage-file`
is passed, and it shadows the jsdom one. You're on 26.3.1, CI pins 24.x, I happened to
be on 24.16.0 — so my run was green and yours red for the same code, and neither of us
was seeing the whole thing. `whats-new.spec.ts` and `plotterext.embedding-host.spec.ts`
being the only specs under `src/` that touch `localStorage` directly is exactly the
footprint you'd predict.

So: master is fine, your branches are fine, and your counts were honest all along. Sorry
for the correction that wasn't one.

**Thanks for going at the CodeRabbit learning directly.** My wrong diagnosis had been
recorded as a project rule, which is worse than no rule at all — it would have taught
every future review to wave away genuine `test:ci` failures as an invocation error. I
see it has acknowledged and replaced it with your version.

**I've written the lesson up** rather than making you spend another round-trip on it —
#688, credited to you. It goes in `DEV-LESSONS-LEARNED.md` under the testing phase with
the version boundary, the error signature and the `--localstorage-file` workaround, and
it sits deliberately next to the bare-vitest entry with a note on how to tell the two
apart, since conflating them is what went wrong here. Node 26 is what somebody gets
today by installing "latest", and the failure is doubly misleading — two
unrelated-looking specs, and an error that reads like a broken test rather than a
runtime quietly shadowing a global. Next person should find it in one search instead of
an afternoon.

**Comment by mairas** [comment]
Both findings from the latest review round were valid and both land here, since
the code is this PR's. Rebased onto #678.

**1. `truncateZoomToDisplay` rounded before truncating** (`6637c467`). Taken.
`toFixed(6)` settled the product by rounding, so any zoom within 5e-7 below a
tenth was lifted onto it: `13.99999995` read as `Zm: 14.0` while the view was
below 14. That is the same defect the truncation was introduced to fix — the
readout naming a level the map has not reached — surviving at a width nobody
would ever see, under a docstring claiming it was gone.

Fixed with one ulp of the product rather than a decimal count, which covers the
`12.3 * 10 === 122.99999999999999` case it was there for and nothing wider.
Verified across both:

```
        current  proposed
12.3       12.3      12.3
13.9757    13.9      13.9
13.99999995  14      13.9
27.999999999 28      27.9
```

The test pins the near-boundary values rather than the mechanism, so it fails
again for any allowance sized in decimals. Worth noting the invariant you asked
for still holds either way — a readout of `14.0` implies zoom ≥ 14, which is
past the layer's `displayMinZoom - 1e-6` bound.

**2. The stale palette close** (`b1a8536f`) — the mechanism and the reasoning
are on #678; this commit applies it to the minimum-zoom palette.

The part worth recording here is why the hand-over key is `<setting>:<chart
id>` and not the chart: an image-adjustment palette opening for the same chart
does not touch the bound, so it is *not* a continuation of that edit and the
bound must still revert. A per-chart session — which is what the finding
proposed — would have suppressed that revert and left a preview the user
cancelled sitting on the map. Both palettes only exist together on this branch,
so that test lives here.

**Verification.** `format:check` clean, `build:web` clean, `test:ci` 566
passing, 0 failing.

**Comment by joelkoz** [comment]
Correcting myself on two things, and then one small ask.

**You'd already done both fixes when I said I was taking them.** You pushed them at
19:46; my comment on #680 went out about twenty minutes later saying "don't start on the
two fixes — I've taken them" and pointing at issue #690. I had current state on #680 and
assumed it covered this PR without re-checking. Sorry — that's exactly the kind of
comment that wastes a contributor's evening, and it was avoidable.

#690 is closed. Nothing in it needs doing.

**#678 merged at `e8a11ddf`**, so your palette hand-over mechanism went in with it rather
than being left behind. That was luck rather than judgement on my part — I merged without
realising the fix was on the branch.

For the record, both fixes are better than what I asked for:

- The zoom allowance is **one ulp of the actual product** rather than the relative epsilon
  I suggested. That covers `12.3 * 10 === 122.99999999999999` and nothing a view can
  actually be at, and the docstring now says why a decimal count was the wrong shape —
  which is the part that stops it regressing.
- Keying the hand-over `<setting>:<chartId>` rather than by chart is the detail I would
  have missed. An image-adjustment palette opening on a chart has no business claiming
  that chart's minimum-zoom edit, and nothing in the review said so.

**The ask: this branch needs a rebase.** #678 was squash-merged, so master now carries its
four commits as the single `71168c06` while this branch still has the originals — same
content, different SHAs, hence the conflict. Dropping them and replaying your own work on
top clears it:

```sh
git fetch upstream
git rebase --onto upstream/master e8a11ddf
git push -f
```

`e8a11ddf` is the boundary — everything from `a9d00c33` up is yours and replays cleanly.
I ran it locally to check before asking: 14 commits, no conflicts, and the resulting
source tree is byte-identical to your current head apart from `DEV-LESSONS-LEARNED.md`
picking up #688 from master.

I could have pushed that to your branch myself — `maintainerCanModify` is on — but you're
clearly mid-session and two people force-pushing the same branch is how work gets lost.
Say the word if you'd rather I did.

Once it's green I'll merge this and then #680.

**Comment by mairas** [comment]
Rebased — `bc8f7a67`. Fourteen commits onto `origin/master`, `#678`'s dropped as
the squash, no conflicts, exactly as you measured. No apology needed for the
overlap; the fixes were worth doing on their merits and the review round is what
found them.

I did not take your `DEV-LESSONS-LEARNED.md` note for granted — the Node 26
entry from #688 came in with the rebase and reads correctly against what I
measured here.

**Verification on the rebased head.** `format:check` clean, `build:web` clean,
`test:ci` 566 passing, 0 failing — on Node 26 with
`NODE_OPTIONS=--localstorage-file=…`, per the lesson.

#680 is rebased on top of this and pushed too (`6f294e56`), so it is ready to
follow this one in without a second round trip.
